### PR TITLE
Feature/cache-locks

### LINKED ISSUE #264: Cache Locks
Currently Kong uses an in-memory cache to avoid making request to the database every time an entity is requested by the system. 

This might lead to a performance loss in a very specific edge-case when lots of incoming requests are being sent to Kong, that triggers simultaneous cache misses which in turn triggers multiple requests to the database, basically DDosing the database and slowing down Kong itself (also called the **Dogpile effect**). 

One of the many examples where this happens is when lots of connections are being sent to an API that has never been loaded into memory before.

The cache client should leverage locks to make sure only one request is being sent to the database everytime, possibly using this module: https://github.com/openresty/lua-resty-lock

This might also be related to #15.

### PR DESCRIPTION
**Pull request by thefosk**
When requesting a specific entity from the database, only one request to the database per node is allowed. This dramatically improves reliability during heavy load.

For example: let's say that 1000 req/s are being processed by a Kong node, and suddenly an in-memory entity is being invalidated and Kong needs to read it again from the database. Before this PR Kong would open 1000req/s to the database until that entity is stored into memory again. With this PR, only one read request will be opened to the database (per node), and the other requests (on the same node) will just wait until it's finally in memory again.

This also helps the performance of new nodes that are being added to the cluster (whose load goes from 0 to 100 real quick).

Closes #264.

### CODE REVIEW

File: kong/tools/database_cache.lua [hunk 1]
```diff
@@ -120,21 +121,42 @@ function _M.all_apis_by_dict_key()
 end
 
 function _M.get_or_set(key, cb)
+  local lock = resty_lock:new("cache_locks", {
+    exptime = 10,
+    timeout = 10
+  })
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: shouldn't the lock be create on the module level instead of on each call to `get_or_set`?

Comment (Tieske) [inline kong/tools/database_cache.lua]: nevermind, the lock can only lock 1 key at a time. So function level lock is required.

File: kong/tools/database_cache.lua [hunk 2]
```diff
@@ -120,21 +121,42 @@ function _M.all_apis_by_dict_key()
...
+
   local value, err
-  -- Try to get
+
+  -- Try to get the value from the cache
+  value = _M.get(key)
+  if value then return value end
+
+  -- The value is missing, aquire a lock
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: typo: acquire

File: kong/tools/database_cache.lua [hunk 3]
```diff
@@ -120,21 +121,42 @@ function _M.all_apis_by_dict_key()
...
+  local elapsed, err = lock:lock(key)
+  if not elapsed then
+    ngx.log(ngx.ERR, "failed to acquire cache lock: "..err)
+  end
+
+  -- Lock acquired. Since in the meanwhile another worker may have
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: typo; meantime

**Comment by mashapedeployment** [comment]
Fixed typos.

**Comment by thibaultCha** [comment]
to be reopened against next

**Pull request closed without merging**